pkg/controller/resourcequota/resource_quota_monitor.go

```diff
@@ -0,0 +1,341 @@
+		return shared.Informer().GetController(), nil
+	}
+	glog.V(4).Infof("QuotaMonitor unable to use a shared informer for resource %q: %v", resource.String(), err)
+
+	// TODO: if we can share storage with garbage collector, it may make sense to support other resources
+	// until that time, aggregated api servers will have to run their own controller to reconcile their own quota.
+	return nil, fmt.Errorf("unable to monitor quota for resource %q", resource.String())
+}
+
+// syncMonitors rebuilds the monitor set according to the supplied resources,
+// creating or deleting monitors as necessary. It will return any error
+// encountered, but will make an attempt to create a monitor for each resource
+// instead of immediately exiting on an error. It may be called before or after
+// Run. Monitors are NOT started as part of the sync. To ensure all existing
+// monitors are started, call startMonitors.
+func (qm *QuotaMonitor) syncMonitors(resources map[schema.GroupVersionResource]struct{}) error {
+	qm.monitorLock.Lock()
+	defer qm.monitorLock.Unlock()
+
+	toRemove := qm.monitors
+	if toRemove == nil {
+		toRemove = monitors{}
+	}
+	current := monitors{}
+	errs := []error{}
+	kept := 0
+	added := 0
+	for resource := range resources {
+		if _, ok := qm.ignoredResources[resource.GroupResource()]; ok {
+			continue
+		}
+		if m, ok := toRemove[resource]; ok {
+			current[resource] = m
+			delete(toRemove, resource)
+			kept++
+			continue
+		}
+		c, err := qm.controllerFor(resource)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("couldn't start monitor for resource %q: %v", resource, err))
+			continue
+		}
+
+		// check if we need to create an evaluator for this resource (if none previously registered)
+		evaluator := qm.registry.Get(resource.GroupResource())
+		if evaluator == nil {
+			listerFunc := generic.ListerFuncForResourceFunc(qm.informerFactory.ForResource)
+			listResourceFunc := generic.ListResourceUsingListerFunc(listerFunc, resource)
+			evaluator = generic.NewObjectCountEvaluator(false, resource.GroupResource(), listResourceFunc, "")
+			qm.registry.Add(evaluator)
+			glog.Infof("QuotaMonitor created object count evaluator for %s", resource.GroupResource())
+		}
+
+		// track the monitor
+		current[resource] = &monitor{controller: c}
+		added++
+	}
+	qm.monitors = current
+
+	for _, monitor := range toRemove {
+		if monitor.stopCh != nil {
+			close(monitor.stopCh)
+		}
+	}
+
+	glog.V(4).Infof("quota synced monitors; added %d, kept %d, removed %d", added, kept, len(toRemove))
+	// NewAggregate returns nil if errs is 0-length
+	return utilerrors.NewAggregate(errs)
+}
+
+// startMonitors ensures the current set of monitors are running. Any newly
+// started monitors will also cause shared informers to be started.
+//
+// If called before Run, startMonitors does nothing (as there is no stop channel
+// to support monitor/informer execution).
+func (qm *QuotaMonitor) startMonitors() {
```

Comment:
This is looking very similar to the GC code.  Need an issue to collapse the impls?

Reply:
opened https://github.com/kubernetes/kubernetes/issues/54515

agree they are similar on purpose.  i think a follow-on to collapse them on common patterns makes sense.  its basically a multi-dimensional resource monitor.  wouldn't be surprised if other controllers want same in future as well.